### PR TITLE
Support for unsharded parameters in state_dict APIs

### PR DESCRIPTION
**Pull request by RdoubleA**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Credit goes to @ifed-uscd for the original commit, I've adapted it a bit to our latest APIs. Original summary:

"This diff adds functionality to shard the model separately from the vocabulary pruning, which allows us to run training keeping the model in bf16 and the vocab pruning in fp32"

#### Changelog
What are the changes made in this PR?
*

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: torchtune/training/_distributed.py [hunk 1]
```diff
@@ -242,7 +245,9 @@ def gather_cpu_state_dict(
         if sharded_param.is_cpu:
             # Move back to device if offloaded to CPU
             sharded_param = sharded_param.to(device)
-        if isinstance(sharded_param._local_tensor, NF4Tensor):
+        if hasattr(sharded_param, "_local_tensor") and isinstance(
```
Comment (pbontrager) [inline torchtune/training/_distributed.py:248]: If the tensor isn't sharded but it still a NF4Tensor, we still need to upcast the datatype as is done on line 271

File: torchtune/training/_distributed.py [hunk 2]
```diff
@@ -242,7 +245,9 @@ def gather_cpu_state_dict(
...
+            sharded_param._local_tensor, NF4Tensor
+        ):
```
Comment (pbontrager) [inline torchtune/training/_distributed.py:250]: For line 274 (can't comment on that line), what happens if full_tensor() is called on an unsharded parameter? It doesn't crash?

Comment (pbontrager) [reply torchtune/training/_distributed.py:250]: This is an edge case, but what if the unsharded param is not on rank0? Will it get moved to cpu properly? What if it's not even on the same node?

**Review by pbontrager — COMMENTED** [review]
Added some questions around edge cases.

File: torchtune/training/_distributed.py (hunk 1 above)
Comment (RdoubleA) [reply torchtune/training/_distributed.py:248]: so maybe another `elif` in the case that the tensor is unsharded, but still NF4Tensor, and just need to call line 271?

File: torchtune/training/_distributed.py (hunk 2 above)
Comment (RdoubleA) [reply torchtune/training/_distributed.py:250]: oh shoot, forgot to copy that logic over from the internal diff

File: torchtune/training/_distributed.py (hunk 1 above)
Comment (pbontrager) [reply torchtune/training/_distributed.py:248]: I think this would be much cleaner if at the top of the loop we do:
```
if hasattr(...):
	# get full tensor (NF4 or .full_tensor)
if full_param is NF4:
	# upcast
if is_rank_zero:
	# the rest the same
```
This code still seems to assume that the unsharded param is on rank0 which isn't guaranteed

**Review by SLR722 — APPROVED** [review]
Looks good to me
